Editor/Scripts/CodeGenerators/RGActionCodeGenerator.cs

```diff
@@ -0,0 +1,111 @@
+            {
+                string scriptText = File.ReadAllText(csFilePath);
+
+                SyntaxTree syntaxTree = CSharpSyntaxTree.ParseText(scriptText);
+
+                var compilation = CSharpCompilation.Create("RGCompilation")
+                    .AddReferences(MetadataReference.CreateFromFile(typeof(object).Assembly.Location))
+                    .AddSyntaxTrees(syntaxTree);
+
+                CompilationUnitSyntax root = syntaxTree.GetCompilationUnitRoot();
+
+                var botActionMethods = root.DescendantNodes()
+                    .OfType<MethodDeclarationSyntax>()
+                    .Where(method =>
+                        method.AttributeLists.Any(attrList =>
+                            attrList.Attributes.Any(attr =>
+                                attr.Name.ToString() == "RGAction")))
+                    .ToList();
+
+                foreach (var method in botActionMethods)
+                {
+                    string className = method.Ancestors().OfType<ClassDeclarationSyntax>().First().Identifier.ValueText;
+                    string methodName = method.Identifier.ValueText;
+
+                    string actionName = methodName;
+                    var actionAttribute = method.AttributeLists.SelectMany(attrList => attrList.Attributes)
+                        .FirstOrDefault(attr => attr.Name.ToString() == "RGAction");
+
+                    if (actionAttribute != null)
+                    {
+                        var attributeArgument = actionAttribute.ArgumentList?.Arguments.FirstOrDefault();
+                        if (attributeArgument != null &&
+                            attributeArgument.Expression is LiteralExpressionSyntax literal)
+                        {
+                            actionName = literal.Token.ValueText;
+                        }
+                    }
+
+                    var semanticModel = compilation.GetSemanticModel(syntaxTree);
+                    var parameterList = method.ParameterList.Parameters.Select(parameter =>
+                        new RGParameterInfo
+                        {
+                            Name = parameter.Identifier.ValueText,
+                            Type = RemoveGlobalPrefix(semanticModel.GetTypeInfo(parameter.Type).Type
+                                .ToDisplayString(SymbolDisplayFormat.FullyQualifiedFormat))
+                        }).ToList();
+
+                    botActionList.Add(new RGActionInfo
+                    {
+                        Object = className,
+                        MethodName = methodName,
+                        ActionName = actionName,
+                        Parameters = parameterList
+                    });
+                }
+            }
+
+            string jsonResult =
+                JsonConvert.SerializeObject(new RGActionsInfo { BotActions = botActionList }, Formatting.Indented);
+            
+            // TODO: send json result to server for typedef generation
```

Comment:
TODOs need to refer to a linear task #

Reply:
Sorry @rcornwall-reg yeah we may have not covered this - for any TODO, we make a linear task with more info about the TODO and then link it in, i.e.

```
// TODO(REG-1234): send json result to server for typedef generation
```